design/002-secret-generator-group.md

```diff
@@ -0,0 +1,277 @@
+
+## Motivation
+
+The motivation is to boost adoption of zero-trust secrets management
+and to make it simple to create secure secrets for machine users
+of services running inside clusters (i.e. databases etc).
+
+### Goals
+
+* enable zero-trust secret management for purely technical and internal users
+* enable secure passwords by default
+* reduce error rates when creating RSA or ECDSA private keys
+
+### Non-Goals
+
+* auto-rotation of secrets
+
+## Proposal
+
+Implement several generators to support not only a number of data types but also generators that interface with external APIs.
+
+### API
+
+We add a new resource group `generator.external-secrets.io`. For each generator we add a new CRD, e.g. for password generation, VaultPKI, ECR, GCR etc.
+This custom resource contains all the configuration details to create a secure value.
+It can be referenced from a `Kind=ExternalSecret` or `Kind=PushSecret`. **or embedded** into it.
+
+The controller generates the value and passes it onto the `Kind=Secret` with all the values
+that have been fetched from the provider.
+When used with a `Kind=PushSecret` it will generate the value and push it to the configured provider.
+
+A generator generates a known set of outputs, e.g. `username` and `password`, `private_key` and `certificate` or others.
+
+#### Example CRD
+
+The following custom resource `Kind=VaultPKIBackend` contains all the parameters for issuing a certificate via HashiCorp Vault PKI engine.
+
+```yaml
+apiVersion: generator.external-secrets.io/v1alpha1
+kind: VaultPKIBackend
+metadata:
+  name: my-vault-pki-creds
+spec:
+  # specific params for generating
+  commonName: example.com
+  altNames:
+  - example.org
+  - localhost
+  # ... more opts
+  auth: {} # service account etc. etc.
+```
+
+This resource can be **embedded or referenced** from an ExternalSecret or PushSecret.
+The ExternalSecret controller generates the secret value and it can be post-processed with
+existing features like template or key-rewrite.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  # ...
+  dataFrom:
+
+    # embedded
+    - generator:
+        apiVersion: generator.external-secrets.io/v1alpha1
+        kind: VaultPKIBackend
+        spec:
+          # same as above
+          commonName: example.com
+          altNames:
+          - example.org
+          - localhost
+          # ... more opts
+          auth: {} # service account etc. etc.issuerRef: "default"
```

Comment:
How would that be handled in the ExternalSecrets CRD? Would it be a `map[string]interface{}` plus some code logic to render if this manifest is a valid CRD? For me, that feels weird, but otherwise I wouldn't know how we could have it dynamically without making ExternalSecrets API a living hell.

Reply:
Second comment, on that: should it generate the actual CRD in the kubernetes namespace? This would allow a lifecycle on the controller logic of the generator per se, where the ExternalSecrets would only be the "Owner" of the generator. This would allow that, if the ES is destroyed, some lifecycle policies could **eventually** apply to delete the generated secret. WDYT?

Reply:
I think the `dataFrom[].generator` should be a [apiextensions.JSON](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types_jsonschema.go#L242) or some type that allows deserialization to a runtime.Object so we can use the controller-runtime tools to deal with that.

> For me, that feels weird, but otherwise I wouldn't know how we could have it dynamically without making ExternalSecrets API a living hell.

We can consider not supporting the inline version but instead force the user to create a dedicated generator resource that can only be used with `generatorRef`. That would be a cleaner implementation. However, If we're considering the `Vault/PKI` use-case that would mean that one would have to create a ExternalSecret+Generator for each and evey TLS certificate :vomiting_face:

> should it generate the actual CRD in the kubernetes namespace?

I think the `Kind=Generator` is just a container for typed configuration values. Conceptually there shouldn't be any cleanup logic needed. It mustn't store any state (maybe by design?!) and thus is reusable across many ExternalSecrets.

Reply:
> I think the dataFrom[].generator should be a [apiextensions.JSON](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apis/apiextensions/v1/types_jsonschema.go#L242) or some type that allows deserialization to a runtime.Object so we can use the controller-runtime tools to deal with that.

Agree with the beneficts of having controller-runtime on our side. Then we can map the object to an equivalent CRD struct definition (and fail if it doesn't exist :smile:).

> That would be a cleaner implementation. However, If we're considering the Vault/PKI use-case that would mean that one would have to create a ExternalSecret+Generator for each and evey TLS certificate vomiting_face

They would have to anyways - the difference would be if the generator is inline or not :cry: .

> I think the Kind=Generator is just a container for typed configuration values. Conceptually there shouldn't be any cleanup logic needed. It mustn't store any state (maybe by design?!) and thus is reusable across many ExternalSecrets.

I agree on that. @knelasevero WDYT? If we all agree I believe it is worth to put this in the design docs itself.